drivers/serial/uart_intel_lw.c

```diff
@@ -0,0 +1,999 @@
+	void *cb_data;  /**< Callback function arg */
+#ifdef CONFIG_UART_INTEL_LW_EOP
+	uint8_t set_eop_cb;
+	uart_irq_callback_user_data_t cb_eop;  /**< Callback function pointer */
+	void *cb_data_eop;  /**< Callback function arg */
+#endif /* CONFIG_UART_INTEL_LW_EOP */
+	uint32_t control_val; /* stores value to set control register. */
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+/*
+ * device config:
+ * stores data that cannot be changed during run time.
+ */
+struct uart_intel_lw_device_config {
+	mm_reg_t base;
+	uint32_t flags;              /* refer to configuration flags */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t  irq_config_func;
+	unsigned int irq_num;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+/**
+ * function prototypes
+ */
+static int uart_intel_lw_irq_update(const struct device *dev);
+static int uart_intel_lw_irq_tx_ready(const struct device *dev);
+static int uart_intel_lw_irq_rx_ready(const struct device *dev);
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+/**
+ * @brief Poll the device for input.
+ *
+ * This is a non-blocking function.
+ *
+ * This driver supports interrupt driven api.
+ * Polling for data under normal operation, might cause unexpected behaviour.
+ * If users wish to poll for data, please use the api:
+ * uart_drv_cmd with CMD_POLL_ASSERT_RTS before polling in.
+ * Then use CMD_POLL_DEASSERT_RTS to resume normal operation after polling.
+ *
+ * @param dev UART device instance
+ * @param p_char Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if input buffer is empty.
+ * -EINVAL if p_char is null pointer
+ */
+static int uart_intel_lw_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	const struct uart_intel_lw_device_config *config = dev->config;
+	struct uart_intel_lw_device_data *data = dev->data;
+	int ret_val = -1;
+	uint32_t status;
+
+	/* generate fatal error if CONFIG_ASSERT is enabled. */
+	__ASSERT(p_char != NULL, "p_char is null pointer!");
+
+	/* Stop, if p_char is null pointer */
+	if (p_char == NULL) {
+		return -EINVAL;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	/* check if received character is ready.*/
+	status = sys_read32(config->base + INTEL_LW_UART_STATUS_REG_OFFSET);
+	if (status & INTEL_LW_UART_STATUS_RRDY_MSK) {
+		/* got a character */
+		*p_char = sys_read32(config->base + INTEL_LW_UART_RXDATA_REG_OFFSET);
+		ret_val = 0;
+	}
+
+	k_spin_unlock(&data->lock, key);
+
+	return ret_val;
+}
+
+/**
+ * @brief Output a character in polled mode.
+ *
+ * This function will block until transmitter is ready.
+ * Then, a character will be transmitted.
+ *
+ * This driver supports interrupt driven api.
+ * Polling for data under normal operation, might cause unexpected behaviour.
+ * If users wish to poll for data, please use the api:
+ * uart_drv_cmd with CMD_POLL_ASSERT_RTS before polling out.
+ * Then use CMD_POLL_DEASSERT_RTS to resume normal operation after polling.
```

Comment:
If you are using RTS/CTS, you will need to check for CTS before sending anything over.

Reply:
This is checked for interrupt driven API.
The check is also redundant, as the IP will not shift out or shift in any data, if CTS is deasserted.